### ISSUE TITLE
High occurrences of "Transporter transfer failed."

### ISSUE THREAD

**Issue by SidiBecker** [bitrise-steplib/steps-deploy-to-itunesconnect-deliver#125]
### Troubleshooting

- [X] I've searched [discuss.bitrise.io](https://discuss.bitrise.io/) for possible solutions.
- Which version of the step is effected? 2.21
- Is the issue reproducible with the latest version? YES
- Does the issue happen sporadically, or every time? SPORADIC 
- Is the issue reproducible locally by following our [local debug guide](https://devcenter.bitrise.io/troubleshooting/debugging-your-build-on-your-own-machine/)? NOT APPLICABLE

### Issue description

Periodically the error occurs: ` Transporter transfer failed.` / `Exception occurred: issuerId is required.`. 
On average, 1 in 10 builds this error occurs, but sometimes the error occurs a few times in a row before it works.

### Bitrise info

- Build URL: https://app.bitrise.io/build/e9980366-c0ae-4ecc-ab81-f75c8f3ceb30
- Bitrise Support enabled: YES
- Log: 

<details>
  <pre>

INFO[14:15:48]  * [OK] Step dependency (go) installed, available. 
Config:
- IpaPath: /Users/vagrant/deploy/Cliente.ipa
- PkgPath: 
- BitriseConnection: off
- ItunesConnectUser: 
- Password: 
- AppPassword: 
- APIKeyPath: *****
- APIIssuer: 69a6de82-****-47e3-************
- AppID: 
- BundleID: [BundleID]
- SubmitForReview: no
- SkipMetadata: yes
- SkipScreenshots: yes
- SkipAppVersionUpdate: yes
- TeamID: [team id]
- TeamName: 
- Platform: ios
- Options: 
- GemfilePath: ./Gemfile
- FastlaneVersion: latest-stable
- ITMSParameters: 
- VerboseLog: true
- BuildURL: https://app.bitrise.io/build/e9980366-c0ae-4ecc-ab81-f75c8f3ceb30
- BuildAPIToken: *****
Inputs with API key authentication found
Setup
fastlane version defined: latest-stable, installing...
$ gem "install" "fastlane" "--no-document"
$ rbenv "rehash"
$ fastlane "-v"
fastlane installation at path:
/Users/vagrant/.rbenv/versions/2.6.9/lib/ruby/gems/2.6.0/gems/fastlane-2.210.1/bin/fastlane
-----------------------------
fastlane 2.210.1
Setup took 6.437116 seconds to complete
Deploy
$ fastlane "deliver" "--api_key_path" "/var/folders/11/nh0v1jld7zd7b9zqm1774gtm0000gn/T/apiKey1842777621/api_key.json" "--precheck_include_in_app_purchases" "false" "--app_identifier" "" "--team_id" "teamid" "--ipa" "/var/folders/11/nh0v1jld7zd7b9zqm1774gtm0000gn/T/ipaOrPkg2536522255/tmp.ipa" "--skip_screenshots" "--skip_metadata" "--skip_app_version_update" "--force" "--platform" "ios"
[14:15:59]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[14:16:00]: Creating authorization token for App Store Connect API
+--------------------------------------+--------------------------------------------------------------------------------+
|                                                deliver 2.210.1 Summary                                                |
+--------------------------------------+--------------------------------------------------------------------------------+
| api_key_path                         | /var/folders/11/nh0v1jld7zd7b9zqm1774gtm0000gn/T/apiKey1842777621/api_key.json |
| precheck_include_in_app_purchases    | false                                                                          |
| app_identifier                       | app                                              |
| team_id                              | teamid|
| ipa                                  | /var/folders/11/nh0v1jld7zd7b9zqm1774gtm0000gn/T/ipaOrPkg2536522255/tmp.ipa    |
| skip_screenshots                     | true                                                                           |
| skip_metadata                        | true                                                                           |
| skip_app_version_update              | true                                                                           |
| force                                | true                                                                           |
| platform                             | ios                                                                            |
| screenshots_path                     | ./screenshots                                                                  |
| metadata_path                        | ./metadata                                                                     |
| app_version                          | 1.361                                                                          |
| edit_live                            | false                                                                          |
| use_live_version                     | false                                                                          |
| skip_binary_upload                   | false                                                                          |
| overwrite_screenshots                | false                                                                          |
| sync_screenshots                     | false                                                                          |
| submit_for_review                    | false                                                                          |
| verify_only                          | false                                                                          |
| reject_if_possible                   | false                                                                          |
| phased_release                       | false                                                                          |
| reset_ratings                        | false                                                                          |
| run_precheck_before_submit           | true                                                                           |
| precheck_default_rule_level          | warn                                                                           |
| ignore_language_directory_validation | false                                                                          |
+--------------------------------------+--------------------------------------------------------------------------------+
[14:16:02]: Uploading binary to App Store Connect
[14:16:03]: Going to upload updated app to App Store Connect
[14:16:03]: This might take a few minutes. Please don't interrupt the script.
[14:16:26]: Transporter transfer failed.
[14:16:26]: 
[14:16:26]: There was a general exception while executing
An exception has occurred: issuerId is required
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at com.apple.transporter.bootstrap.BootstrapperPhase1.bootstrap(BootstrapperPhase1.java:107)
[14:16:26]: [iTMSTransporter] 	at com.apple.transporter.bootstrap.BootstrapperPhase1.bootstrap(BootstrapperPhase1.java:78)
[14:16:26]: [iTMSTransporter] 	at com.apple.transporter.launcher.Launcher.launchBootstrapper(Launcher.java:37)
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
[14:16:26]: [iTMSTransporter] 	at com.apple.transporter.launcher.Application.start(Application.java:450)
[14:16:26]: [iTMSTransporter] 	at com.apple.transporter.launcher.Application.main(Application.java:947)
[14:16:26]: [iTMSTransporter] 
[14:16:26]: [iTMSTransporter] [2022-10-19 14:16:26 GMT] <main> DBG-X: Returning 1
[14:16:26]: iTunes Transporter output above ^
[14:16:26]: There was a general exception while executing
An exception has occurred: issuerId is required
Return status of iTunes Transporter was 1: There was a general exception while executing
\nAn exception has occurred: issuerId is required
The call to the iTMSTransporter completed with a non-zero exit status: 1. This indicates a failure.
  </pre>
</details>

### Steps to reproduce

1. To perform archiving, export and deploy, we use API Key, Issuer and Key ID. Our .p8 file is in Code Sigining -> Generic File Storage. We have multi apple account setup.

yml setup:

```yml
- xcode-archive@4.7:
        inputs:
        - scheme: "$PROJECT_NAME"
        - configuration: ''
        - team_id: [team id]
        - compile_bitcode: 'no'
        - custom_export_options_plist_content: ''
        - xcodebuild_options: ''
        - force_provisioning_profile_specifier: ''
        - export_method: app-store
        - verbose_log: 'yes'
        - export_all_dsyms: 'no'
        - distribution_method: app-store
        - automatic_code_signing: api-key
        - export_development_team: "$IOS_TEAM_RELEASE"
        - api_key_path: "$URL_P8"
        - api_key_id: "$KEY_ID"
        - api_key_issuer_id: "$ISSUER_ID"
        - project_path: "$BITRISE_SOURCE_DIR/platforms/ios/$PROJECT_NAME.xcworkspace"

   - deploy-to-itunesconnect-deliver@2.21:
        inputs:
        - bundle_id: "$PACKAGE_NAME"
        - skip_app_version_update: 'yes'
        - verbose_log: 'yes'
        - connection: 'off'
        - api_key_path: "$URL_P8"
        - team_id: "$IOS_TEAM_RELEASE"
        - api_issuer: "$ISSUER_ID"
```

**Comment by k-thorat** [comment]
We are seeing this as well. Does this #126 fixes the issue?

**Comment by mateusz1913** [comment]
👋 I also encountered this issue and I can confirm that after pointing to the #126 PR branch, the build was successfully submitted to App Store Connect

```yaml
_deploy-ios-app-store:
    steps:
    - git::https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver.git@SWAT-788-itms3-fix:
        inputs:
        - bundle_id: $IOS_BUNDLE_ID
```

**Comment by SidiBecker** [comment]
I'm successfully submitting apps with changes [#123](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/pull/126) too

**Comment by ludwig-pro** [comment]
Same issue for us, I will try the fix #126 monday, thanks @mateusz1913

**Comment by SidiBecker** [comment]
The version [2.21.5](https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver/releases/tag/2.21.5) just release. 
I'll test it today and close this issue if all works.

**Comment by lpusok** [comment]
Released new version 2.21.5 with fix included.

@k-thorat, @mateusz1913, @SidiBecker, @arnoldc, @siddheshbhasme, @rpadilla6, @lisamartin00, @Cameronese11, @mendozamiche, @LucasClaude, @pretorh and @ludwig-pro: [CTA]
Please remove usages `git::https://github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver.git@SWAT-788-itms3-fix:` of as it the branch will be deleted. Use just `deploy-to-itunesconnect-deliver`. Thanks.

**Comment by SidiBecker** [comment]
@lpusok Is already available in bitrise? There the last version still is 2.21.4

**Comment by lpusok** [comment]
@SidiBecker not yet, should be available shortly. Used a manual trigger as the automation did not finish.

**Comment by SidiBecker** [comment]
Available now.

**Comment by ludwig-pro** [comment]
I'm testing this now thanks

**Comment by ludwig-pro** [comment]
don't work for us :/ 

```
Deploy
Failed to read Xcode version: %!w(*errors.errorString=&{failed to parse xcodebuild version output (objc[24255]: Class AppleTypeCRetimerRestoreInfoHelper is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa3620) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d8598). One of the two will be used. Which one is undefined.
objc[24255]: Class AppleTypeCRetimerFirmwareAggregateRequestCreator is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa3670) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d85e8). One of the two will be used. Which one is undefined.
objc[24255]: Class AppleTypeCRetimerFirmwareRequestCreator is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa36c0) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d8638). One of the two will be used. Which one is undefined.
objc[24255]: Class ATCRTRestoreInfoFTABFile is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa3710) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d8688). One of the two will be used. Which one is undefined.
objc[24255]: Class AppleTypeCRetimerFirmwareCopier is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa3760) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d86d8). One of the two will be used. Which one is undefined.
objc[24255]: Class ATCRTRestoreInfoFTABSubfile is implemented in both /usr/lib/libauthinstall.dylib (0x1e6aa37b0) and /Library/Apple/System/Library/PrivateFrameworks/MobileDevice.framework/Versions/A/MobileDevice (0x1054d8728). One of the two will be used. Which one is undefined.
Xcode 13.0
Build version 13A233)})
|                                                                              |
+---+---------------------------------------------------------------+----------+
| x | deploy-to-itunesconnect-deliver@2 (exit code: 1)              | 7.58 sec |
+---+---------------------------------------------------------------+----------+
| Issue tracker: ...trise-steplib/steps-deploy-to-itunesconnect-deliver/issues |
| Source: .../github.com/bitrise-steplib/steps-deploy-to-itunesconnect-deliver |
+---+---------------------------------------------------------------+----------+
    
```

**Comment by SidiBecker** [comment]
Archive on 4.7.0 and last version deploy works fine for me. The 4.6.0 version has some updates of authentication and signing. Could you try updating the archive step?

**Comment by arnoldc** [comment]
can confirm its working also in my end using
Xcode Archive (4.7) + App Store Connect with Deliver  (2.21.5)
thank you

**Comment by lpusok** [comment]
@ludwig-pro can you share a build URL please?

**Comment by ludwig-pro** [comment]
@lpusok https://app.bitrise.io/build/4a7f15cb-bdfe-448e-894f-38bb84b81269

thanks for the support
I will also upgrade Xcode archive 4.7

**Comment by ludwig-pro** [comment]
I retry the build with App Store Connect with Deliver `2.21.6`

**Comment by ludwig-pro** [comment]
Fixed with App Store Connect with Deliver `2.21.6` 🚀

**Comment by SidiBecker** [comment]
Closing this issue. Thank you all 🚀
